### ISSUE TITLE
[PAF] Failed to open libretro core when trying to open ROMS. 

### ISSUE THREAD

**Issue by BetaLeaf** [libretro/RetroArch#5074]
## Description

RetroArch will load cores from a different folder than the one I specified.

### Expected behavior

RetroArch should load the bit specific core for the frontend I'm using atm.

### Actual behavior

I get a "failed to open libretro core" error.

### Steps to reproduce the bug 

(Download from https://cloud.betaleaf.net/s/lNBIF87ECbHjKtq to skip the setup steps 1-8 & 12.)

1. Download and extract both 32 and 64 bit versions into separate folders which we will call app32 and app64 from now on.
2. Move the retroarch.cfg from either app32 or app 64 to another separate folder which we will call the data folder from now on.
3. Remove the following lines from the retroarch.cfg in the data folder: 
(Just remove the lines if they exist, even if they do not match exactly as below.)

> core_updater_buildbot_url = ""
libretro_directory = ""

4. Add or replace the following lines from the retroarch.cfg in the data folder: 
(Replace "%PAL:DataDir%" with the path to the data folder.)

> dpi_override_value = "160"
input_joypad_driver = "xinput"
input_osk_overlay_enable = "false"
load_dummy_on_core_shutdown = "false"
menu_collapse_subgroups_enable = "true"
config_save_on_exit = "false"
video_driver = "gl"
system_directory = "%PAL:DataDir%\system"
cheat_database_path = "%PAL:DataDir%\cheats"
content_history_path = ":\content_history.lpl"
content_music_history_path = ":\content_music_history.lpl"
content_video_history_path = ":\content_video_history.lpl"
content_image_history_path = ":\content_image_history.lpl"
input_remapping_directory = "%PAL:DataDir%\config\remaps"
thumbnails_directory = "%PAL:DataDir%\thumbnails"
playlist_directory = "%PAL:DataDir%\playlists"
screenshot_directory = "%PAL:DataDir%\screenshots"
core_options_path = "%PAL:DataDir%\retroarch-core-options.cfg"

5. Inside app32, create a new retroarch.cfg and insert the following lines: 
(Replace "%app32%" with the path to the app32 folder.)

> core_updater_buildbot_url = "http://buildbot.libretro.com/nightly/win-x86/latest/"
libretro_directory = "%app32%\cores"

6. Inside app64, create a new retroarch.cfg and insert the following lines: 
(Replace "%app64%" with the path to the app64 folder.)

> core_updater_buildbot_url = "http://buildbot.libretro.com/nightly/win-x86_64/latest/"
libretro_directory = "%app64%\cores"

7. Run RetroArch.exe from the app32 folder with the following command lines: 
(Replace "%PAL:DataDir%" with the path to the data folder.)

> retroarch.exe --appendconfig %PAL:DataDir%\retroarch.cfg --save %PAL:DataDir%\saves --savestate %PAL:DataDir%\states --record %PAL:DataDir%\records

8. Download gambette core and a compatible ROM to test.
9. Search for the ROM file and add it into your playlist.
10. From the Game Boy Color playlist, launch the ROM.
11. Close RetroArch.exe
12. Run RetroArch.exe from the app64 folder with the following command lines: 
(Replace "%PAL:DataDir%" with the path to the data folder.)

> retroarch.exe --appendconfig %PAL:DataDir%\retroarch.cfg --save %PAL:DataDir%\saves --savestate %PAL:DataDir%\states --record %PAL:DataDir%\records
13. Repeat steps 8 - 11.

You should have gotten a "failed to open libretro core" error by now. This is what I'm trying to fix.
This also happens if I try to load the ROM from the content history playlist after doing steps 1-13.
This also happens if I try to load the core first, then load the ROM. However, sometimes I can get it to work. I haven't exactly pinned this one down yet.
It seems to favor the app32 or app64 version of the core folder, depending on what was loaded first.

### Bisect Results

See above.

### Version/Commit
You can find this information under Information/System Information

- RetroArch: 1.6.0 (Git Version a4358de)

### Environment information

- OS: Windows 10 Home
- Compiler: N/A

### Why?

I want to make a version of RetroArch that is compatible with PAF specifications. I use the PortableApps.com Platform to launch most of my apps on my computer. I don't like having apps installed locally and would prefer to have them portable. I am able to get everything portable except for the cores. Even tho I specific to keep the 32 and 64 bit cores separate and specific to load 32 bit cores when im using 32 bit retroarch and 64 bit cores when im using 64 bit retroarch, it seems to ignore this setting.

**Comment by hizzlekizzle** [comment]
It sounds like they're trying to load cores from the wrong directory. In your retroarch.cfg, is the "libretro_directory" using a relative path? i.e.,
libretro_directory = ":\cores"

**Comment by BetaLeaf** [comment]
### TL;DR.:
No I am using an absolute path. My launcher automatically update this value to ensure it is always the correct path. RetroArch seems to ignore this though.

### Full:
My PAF config files does this command:

Find=%PAL:LastDrive%%PAL:LastPackagePartialDir%
Replace=%PAL:Drive%%PAL:PackagePartialDir%

This means it will, on execution of the launcher, it will check a file called %APPNAME%Settings.ini, where %APPNAME% is obviously the name of the app. In this case, this file is RetroArchPortableSettings.ini and resides in the ```.\Data\settings\``` folder.
It looks a little something like this:
```
[RetroArchPortableSettings]
LastDrive=D:
LastDirectory=\RetroArchPortable
[PortableApps.comLauncherLastRunEnvironment]
PAL:LastPortableAppsBaseDir=D:
```
Lets also say the current environment for ```%PAL:Drive%%PAL:PackagePartialDir%``` is ```G:\PortableApps\RetroArchPortable```
###### It should be noted that LastDirectory is exactly the same as %PAL:PackagePartialDir%

Every time the app is executed, the launcher loads these values and uses them to update data files. On exit, these values are updated with the environment the app just finished running on.

Then, it open the RetroArch.cfg file and search the file for any values that match the query string. In this case, with the above example, after my launcher executes, but before RetroArch is run, it is going to search for ```D:\RetroArchPortable``` and replace it with the current environment ```G:\PortableApps\RetroArchPortable```.

You can see what these variable are suppose to represent from: https://portableapps.com/manuals/PortableApps.comLauncher/ref/envsub.html#ref-envsub

RetroArch doesn't seem to be able to handle this though, despite it being updated to the correct absolute path. 

P.S. I included my environment, if you wanted to download and test it out. Run it and turn on debug logging and try to open a ROM on both 32 bit and 64 bit. You can change which one you use in the Launcher.ini found at ```.\App\AppInfo\Launcher\Launcher.ini```
You can see what each option does in this INI from:
https://portableapps.com/manuals/PortableApps.comLauncher/ref/launcher.ini/index.html#ref-launcher-ini

**Comment by fr500** [comment]
ahh this is pretty cool.
I like the PAF format a lot.

You don't need all that searching and replacing tho.

Anything you prefix with :\ is made relative automatically and expanded by the program at runtime without saving anything.

I'm very interested. Feel free to reopen!

**Comment by BetaLeaf** [comment]
Unfortunately I tried that. Was still using the wrong bit dolls. I had a friend of mine fix it for me. I'll link later

**Comment by BetaLeaf** [comment]
https://yadi.sk/d/QHwBxasU3KmbKM
